src/core/libraries/kernel/sync/semaphore.h

```diff
@@ -0,0 +1,85 @@
+// SPDX-FileCopyrightText: Copyright 2024 shadPS4 Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include <atomic>
+#include <chrono>
+
+#include "common/assert.h"
+#include "common/types.h"
+
+#ifdef _WIN64
+#include <windows.h>
+#else
+#include <semaphore>
+#endif
+
+namespace Libraries::Kernel {
+
+class BinarySemaphore {
+public:
+    BinarySemaphore();
+    ~BinarySemaphore();
+
+    void release();
+    void acquire();
+
+    bool try_acquire();
+    bool try_acquire_for(std::chrono::microseconds timeout);
+
+    template <class Clock, class Duration>
+    bool try_acquire_until(const std::chrono::time_point<Clock, Duration>& abs_time) {
+#ifdef _WIN64
+        auto now = Clock::now();
+        if (now >= abs_time) {
+            return false;
+        } else {
+            auto timeout = std::chrono::duration_cast<std::chrono::microseconds>(abs_time - now);
+            return try_acquire_for(timeout);
+        }
+#else
+        sem.try_acquire_until(abs_time);
+#endif
+    }
+
+private:
+#ifdef _WIN64
+    HANDLE sem;
+#else
+    std::binary_semaphore sem;
+#endif
+};
+
+class CountingSemaphore {
+public:
+    CountingSemaphore(s32 initialCount, s32 maxCount);
+    ~CountingSemaphore();
+
+    void release();
+    void acquire();
+
+    bool try_acquire();
+
+    template <class Clock, class Duration>
+    bool try_acquire_until(const std::chrono::time_point<Clock, Duration>& abs_time) {
+#ifdef _WIN64
+        auto duration = abs_time.time_since_epoch();
+        auto milliseconds = std::chrono::duration_cast<std::chrono::milliseconds>(duration).count();
+
+        u64 res = WaitForSingleObjectEx(sem, static_cast<u64>(milliseconds), true);

```

Comment:
`BinarySemaphore` is passing a duration timeout but here you are just passing the plain epoch timestamp. Shouldn't this be subtracted from `now()`?